lib/socket.c

```diff
@@ -567,100 +567,111 @@ iscsi_read_from_socket(struct iscsi_context *iscsi)
 {
 	struct iscsi_in_pdu *in;
 	ssize_t data_size, count, padding_size;
+	int available;
 
-	if (iscsi->incoming == NULL) {
-		iscsi->incoming = iscsi_szmalloc(iscsi, sizeof(struct iscsi_in_pdu));
-		iscsi->incoming->hdr = iscsi_szmalloc(iscsi, ISCSI_RAW_HEADER_SIZE + ISCSI_DIGEST_SIZE);
-		if (iscsi->incoming == NULL) {
-			iscsi_set_error(iscsi, "Out-of-memory: failed to malloc iscsi_in_pdu");
-			return -1;
-		}
+	/* check how much data is in the input buffer of the socket and read as many
+	 * PDUs as available */
+	if (ioctl(iscsi->fd, FIONREAD, &available) < 0) {
```

Comment:
Not in Linux. In Linux poll sometimes returns POLLIN even if the receive buffer is empty. In this case available would be 0, but read would not return 0 but -1 with errno = EWOULDBLOCK.